### ISSUE TITLE
Provide a clean and simple math module

### ISSUE THREAD

**Issue by torkleyy** [amethyst/amethyst#200]
I think the user shouldn't explicitly use something like `cgmath` and I also think that we are using `Vector3<f32>` and not `f64`, so I'd like a math module with typedefs for commonly used math structures. If you agree I'd start a PR.

**Comment by vitvakatu** [comment]
I agree. Writing `Vector3<f32>` is boring. Something like `Vec3` is simpler, shorter, easier to read and write.

**Comment by Aceeri** [comment]
I'm somewhat hesitant about forcing usage of f32 for all `Vector3` to possibly save a couple keystrokes. Normally I'd expect the compiler to be able to infer this informaton very easily as well.

**Comment by torkleyy** [comment]
Not for parameters, field types and return types. And when do you need an `f64`?

**Comment by kvark** [comment]
I think `Vec3f` is short and self-descriptive.

**Comment by nchashch** [comment]
TL;DR: `f64`s can be useful for exotic physics things (like orbital mechanics).

`f64`s can be useful if you deal with large floating point numbers. For example in my [orbital-mechanics](github.com/nchashch/orbital-mechanics-rust) experiment I could achieve more accuracy using `f64`s. If you use meters for example average earth radius is `6.4e6` meters, `f32` only have 7 decimal places so you get at most `0.1` meter accuracy. It gets worse say for moon orbit, moon is around `3.85e8` meters away from earth center so you would get at most `10.0` meter accuracy. It isn't a big deal if you just want to say render orbits, but if for example you wanted to detect collisions for objects that are smaller than `10.0` meters it would be a problem.

Probably `f64`s would be needed for quite a small number of use cases. We should probably see how costly would it be to make transform/physics system generic over floating point number types and then decide if we want it.

`f64`s have around 15 decimal places so they shoud give at most `1.0e-9` meters accuracy for earth radius and `1.0e-7` meters for the moon orbit radius. I got less than `1.0e-3` meters error with 99% reliability for converting between position/velocity vectors and keplerian orbital elements (if I measured it correctly) with my `orbital-mechanics-rust` experiment using `f64`s, with orbit radius going from `1.0` to `1.0e10` meters for 10000 samples (measured using tests in the src/lib.rs).

**Comment by vitvakatu** [comment]
I suppose we can make it generic and use default type parameters. 
```
LocalTransform<f64> // in case we want better precision 
LocalTransform<f32> // this is default and equal to...
LocalTransform          // ...this line
```

**Comment by Aceeri** [comment]
Would this mainly just be typedefs to cgmath types?

**Comment by torkleyy** [comment]
@Acereri yes I think so.

@nchashch what about conditional compilation? One `typedef Real = f32` with a cfg and you can change it if you need an f64

**Comment by vitvakatu** [comment]
Bad example. I should have shown it using `Vector` type. This issue is about cgmath types only, but it closely related with first statement in #205.

**Comment by Aceeri** [comment]
I like the idea of having `LocalTransform<T>` where it takes in some sort of num type similar to cgmath and then uses that to construct the cgmath types. For instance:
```rust
struct Transform<T: BaseFloat> {
    quaternion: cgmath::Quaternion<T>,
    translation: cgmath::Vector3<T>,
    scale: cgmath::Vector3<T>,
}
```
Then we could also have the physics system take in a generic related to the precision, e.g. `PhysicsSystem<T: BaseFloat>` which would take all the transforms and modify them with a precision of `T` (truncated if the transforms are `f32` though, so if it is a `Transform<f32>` then a `PhysicsSystem<f64>` will only provide `f32` precision for that transform).

I disagree heavily that we should do anything like:
```rust
type Vec3f = cgmath::Vector3<f32>;
type Vec3f64 =cgmath::Vector3<f64>;
...
Since now we lose the advantage of generic precision and it just seems like unnecessary abstraction.
```

**Comment by torkleyy** [comment]
I'm proposing to close this issue because the opinions seem to be pretty split. Also, saying "reexporting makes it easier for the user" is probably subjective. To sum up the pros and cons of using a compile time precision setting:

### Pros
+ Simpler to write (`Vector3<f32>` vs `Vec3f`)
+ Simplifies internal engine code (No additional type parameter for every math related struct)
### Cons
- Implicit, user isn't aware of math library
- Only one precision is allowed for everything, varying the precision for different structs is not possible

Suggested compromise:

* Use type parameters for math / physics structs (concerning `f32` / `f64`)
* But make `f32` default to make it easy for the user

Everyone okay with this proposal?

**Comment by torkleyy** [comment]
Seems we stay with the current system.

Only issue: from time to time people come to gitter and ask how to do some camera transformation or similar things.

**Comment by torkleyy** [comment]
@vitvakatu 

> I'm not sure that we should close #200

Could you please elaborate how you would proceed?

**Comment by OvermindDL1** [comment]
For note, fixed-point would be quite useful as well so do not hard-code everything to just expect floats either if possible?

**Comment by torkleyy** [comment]
@OvermindDL1

The proposed solution is to let the user choose which math library to use, so it pretty much depends on the math library.

cgmath does not support this, though. Also, are fixed point numbers suitable for games?

**Comment by OvermindDL1** [comment]
> Also, are fixed point numbers suitable for games?

Very much, in my C++ project it is a 2D tile based game where each tile is broken up into 16x16 parts, thus I have a fixed point of 28 bits with 4 bits point with the suitable math to work on it.  This ensures that everything is always perfectly aligned in the game as well as a few other niceties.  A lot of the math works on this as well.  I do still have a floating point vector for other work, but the fixed point is very nice as it is an 'infinite' world (up to 24 bits size of tiles, which is more than a player could ever reach, though if they did most of the code 'wraps around' purposefully anyway).

But yes, all positions of all objects are fixed point and no worry about loss of accuracy at any point or so.

**Comment by ebkalderon** [comment]
@torkleyy First, I was under the impression that we would stick to `cgmath` only? And second, I concur that fixed point math is used very often in 2D grid-based games.

**Comment by torkleyy** [comment]
@ebkalderon Okay, I'm fine with both solutions, but IIRC @nchashch wanted to have support for different math libraries.

> Second, I concur that fixed point math is used very often in certain 2D games.

I don't think you could do fixed point math easily, anyway. It's not supported by maths and graphics libraries, @OvermindDL1 could you also work with `f64`s?

**Comment by kvark** [comment]
> I don't think you could do fixed point math easily, anyway. It's not supported by maths and graphics libraries, @OvermindDL1 could you also work with f64s?

Graphics libraries do support a form of fixed point in their normalized formats. Either way, it's somewhat trivial to pass integers into shaders and treat them as fixed point, so I wouldn't take it as a blocker.

Overall, while fixed-point math is incredibly useful in some cases, I don't think Amethyst needs to feature it specifically. Focusing on `cgmath` is good way to proceed, if you need something delivered in our life time ;)

**Comment by OvermindDL1** [comment]
> I don't think you could do fixed point math easily, anyway. It's not supported by maths and graphics libraries, @OvermindDL1 could you also work with f64s?

I 'could', however it is significantly larger in size (I often use shorts for example) and has different performance and especially comparison characteristics, it is horrible to match on them for example in, say, a triemap/hashmap of some sort as one example, any floating point type should absolutely *not* be used for those.

> Overall, while fixed-point math is incredibly useful in some cases, I don't think Amethyst needs to feature it specifically. Focusing on cgmath is good way to proceed, if you need something delivered in our life time ;)

It will be preventing a lot of use-cases, including near every game I've made in the past ten years.  I near always have world positions be either integral or fixed-point especially as it makes for very fast and efficient lookup that cannot be done with floating point (have to use a quadtree or some other partitioning style, which is significantly slower than a hash lookup or an array lookup, usually array in most of mine).  I would have a *substantial* amount of floating point to integral conversions, which is not too good for both readability and performance..

I am curious, if cgmath is such a standard library for this then how does it not support this style already?  The library I've been using in C++ for years is [Eigen](http://eigen.tuxfamily.org), which does have a FixedPoint template specialization available (in addition to SSE and so forth optimizations, views, etc... etc...)

**Comment by torkleyy** [comment]
@OvermindDL1 

Probably you could use fixed point numbers in your grids and just convert them for rendering?

**Comment by OvermindDL1** [comment]
> Probably you could use fixed point numbers in your grids and just convert them for rendering?

I could but the conversion of a (fixed point) integer to float in a tight loop does not sounds very performant compared to a usual memcpy to an OGL array.  Thankfully the tiles themselves, though numerous, are cached in large polyginal chunks, but the entities that move on top, which can also be numerous (many hundreds of thousands are common, over a million is not unheard of, handled very well in C++ with Eigen) sounds amazingly more painful, though it might make me do some tests of offloading physics and AI calculations onto the GPU via OpenCL or so, hmm...

**Comment by kvark** [comment]
Oh wait, did I accidentally close it? Sorry!

**Comment by kvark** [comment]
> I am curious, if cgmath is such a standard library for this then how does it not support this style already? 

There is barely any library in Rust that is mature. Amethyst will go faster if it settles on *something* that might as well be `cgmath`. Let's try adding the fixed-point support in there, should we?

**Comment by torkleyy** [comment]
:+1: for @kvark's solution

Number types should not be a concern of the game engine, but the math library (and the consumer, the graphics library)

**Comment by OvermindDL1** [comment]
> There is barely any library in Rust that is mature. Amethyst will go faster if it settles on something that might as well be cgmath. Let's try adding the fixed-point support in there, should we?

That would be awesome, I'd absolutely adore an Eigen-like library for Rust, though bringing something up to anywhere near feature parity of Eigen (especially its C++ template optimization passes that I'm not even sure how you could do in Rust at this time) would be beyond fantastic!

**Comment by Binero** [comment]
Limiting the amount of math structures that are exposed to the user would likely be a good solution. Where some math absolutely needs to be exposed, like in the Camera, I would prefer common representations like `[f32; 9]` or `mint` types.

**Comment by Xaeroxe** [comment]
+1 for mint types.  We're unable to pick a good set of math types at the moment and mint is a good common ground.

**Comment by minecrawler** [comment]
@Binero Please check #342 and see if the camera representation is to your liking. I use `LocalTransform` and `Orientation` for the (new) mathematical representation of the camera and movement/rotation helpers, which use `[f32; 3]` (for example [here](https://github.com/amethyst/amethyst/pull/342/files#diff-376d750ad2dcd8dbff3ff04643253bceR6)) and cgmath's `into()`/`from()` methods to store and work with the data as needed. cgmath does implement mint, too, so you can go wild with that stuff ;)

Since I am still in the middle of rewriting code, making changes at this point is very simple ;)

**Comment by kvark** [comment]
Directly related - https://github.com/amethyst/amethyst/issues/236#issuecomment-335250813
